File: CS-IOS/Classes/Core/Categories/UIKit/UITableView+CSExtension.m
//
//  Created by Rene Dohan on 6/11/12.
//


#import "CSLang.h"
#import "UIView+CSExtension.h"
#import "NSObject+CSExtension.h"
#import "UIViewController+CSExtension.h"
#import "UINavigationController+CSExtension.h"
#import "UIView+CSDimension.h"

@implementation UITableView (CSExtension)

- (instancetype)setupTable:(id <UITableViewDelegate, UITableViewDataSource>)parent {
    self.delegate = parent;
    self.dataSource = parent;
    [self reloadData];
    return self;
}

- (instancetype)setupTable:(id <UITableViewDelegate>)delegate :(id <UITableViewDataSource>)dataSource {
    self.delegate = delegate;
    self.dataSource = dataSource;
    [self reloadData];
    return self;
}

- (instancetype)setHeader:(UIView *)header {
    self.tableHeaderView = header;
    return self;
}

- (UIView *)setFooter:(UIView *)header {
    self.tableFooterView = header;
    return header;
}

- (void)deselectSelectedRow {
    [self deselectRowAtIndexPath:self.indexPathForSelectedRow animated:YES];
}

- (void)deselectSelectedRows:(BOOL)animated {
    for (NSIndexPath *path in self.indexPathsForSelectedRows)
        [self deselectRowAtIndexPath:path animated:animated];
}

- (UITableViewCell *)getCell:(Class)type {
    var object = [self dequeueReusableCell:[type description]];
    if (!object) object = [type create];
    return object;
}

- (id)dequeueReusableCell:(NSString *)identifier {
    return [self dequeueReusableCellWithIdentifier:identifier];
}

- (UITableViewHeaderFooterView *)getHeaderFooter:(Class)type {
    var object = [self dequeueReusableHeaderFooter:[type description]];
    if (!object) object = [type create];
    return object;
}

- (UITableViewHeaderFooterView *)dequeueReusableHeaderFooter:(NSString *)identifier {
    return [self dequeueReusableHeaderFooterViewWithIdentifier:identifier];
}

- (instancetype)hideEmptyCellSplitterBySettingEmptyFooter {
    [self setTableFooterView:[UIView.alloc initWithFrame:CGRectMake(0, 0, 0, 0.1)]];
    return self;
}

- (instancetype)toggleEditingAnimated:(BOOL)animated {
    [self setEditing:!self.isEditing animated:animated];
    return self;
}

- (instancetype)toggleEditing {
    [self setEditing:!self.isEditing animated:YES];
    return self;
}

- (UITableViewCell *)cellView:(Class)viewClass :(void (^)(UITableViewCell *))onCreate {
    var cell = [self dequeueReusableCell:[viewClass className]];
    if (!cell) invokeWith(onCreate, cell = [self createCell:viewClass]);
    return cell;
}

- (UITableViewCell *)createCell:(Class)viewClass {
    val cell = [UITableViewCell.alloc initWithStyle:UITableViewCellStyleDefault reuseIdentifier:[viewClass className]];
    UIView *view = [viewClass.create construct];
    [cell size:CGSizeMake(self.width, self.rowHeight = view.height)];
    [cell.contentView matchParent];
    [cell.contentView add:view].matchParent;
    view.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    return cell;
}

- (UITableViewCell *)getCellWithStyle:(NSString *)id :(UITableViewCellStyle)style {
    return [self getCellWithStyle:id :style :nil];
}

- (UITableViewCell *)getCellWithStyle:(NSString *const)id :(enum UITableViewCellStyle)style :(void (^)(UITableViewCell *cell))onCreate {
    var cell = [self dequeueReusableCell:id];
    if (!cell) runWith(onCreate, cell = [UITableViewCell.alloc initWithStyle:style reuseIdentifier:id]);
    return cell;
}
@end

File: Renetik/Renetik/Classes/Core/Extensions/UIKit/UITableView+CSExtension.swift
//
// Created by Rene on 2018-11-29.
//

import UIKit
import RenetikObjc

public extension UITableView {

    override open func construct() -> Self {
        super.construct()
        return self
    }

    @discardableResult
    public func reloadKeepSelection() -> Self {
        let paths = indexPathsForSelectedRows
        reload()
        paths?.forEach { path in selectRow(at: path, animated: false, scrollPosition: .none) }
        return self
    }

    public func cell(style: UITableViewCell.CellStyle,
                     onCreate: ((UITableViewCell) -> Void)? = nil) -> UITableViewCell {
        cell(with: "Cell", style: style, onCreate: onCreate)
    }

    public func cell(with identifier: String, style: UITableViewCell.CellStyle,
                     onCreate: ((UITableViewCell) -> Void)? = nil) -> UITableViewCell {
        var cell = dequeueCellWith(identifier: identifier)
        if cell.isNil {
            cell = UITableViewCell(style: style, reuseIdentifier: identifier)
            onCreate?(cell!.construct())
        }
        return cell!
    }

    @discardableResult
    public func register<CellType: UITableViewCell>(cell type: CellType.Type) -> Self {
        register(type, forCellReuseIdentifier: type.className())
        return self
    }

    private func dequeue<CellType: UITableViewCell>(cell type: CellType.Type) -> CellType {
        let cell = dequeueCellWith(identifier: type.className()) as? CellType
        return cell.isNil ? register(cell: type).dequeueCellWith(identifier: type.className()) as! CellType : cell!
    }

    public func dequeue<CellType: UITableViewCell>(
            cell type: CellType.Type, onCreate: ((CellType) -> Void)? = nil) -> CellType {
        var cell = dequeue(cell: type)
        if cell.contentView.isEmpty {
            onCreate?(cell)
            cell.contentView.matchParent()
            cell.width(width, height: width)
            cell.construct()
        }
        return cell
    }

    @discardableResult
    public func toggleEditing(animated: Bool = true) -> Self {
        setEditing(!isEditing, animated: animated)
        return self
    }

    /*
    * This triggers table reload !!!
    */
    @discardableResult
    public func hideEmptySeparatorsByEmptyFooter() -> Self {
        tableFooterView = UIView(frame: CGRect(x: 0, y: 0, width: 0, height: 0.1))
        return self
    }

    public func deselectSelectedRow(animated: Bool = true) {
        indexPathForSelectedRow?.also { path in deselectRow(at: path, animated: animated) }
    }

    public func deselectSelectedRows(animated: Bool = true) {
        for path in indexPathsForSelectedRows ?? [] { deselectRow(at: path, animated: animated) }
    }

    public func dequeueCellWith(identifier: String?) -> UITableViewCell? {
        dequeueReusableCell(withIdentifier: identifier ?? "")
    }

    @discardableResult
    public func reload() -> Self {
        reloadData()
        return self
    }

    @discardableResult
    public func set<View: UIView>(header view: View) -> View {
        tableHeaderView = view
        return view
    }

    @discardableResult
    public func set(footer view: UIView?) -> UIView? {
        tableFooterView = view
        return view
    }

    @discardableResult
    public func delegates(_ delegates: UITableViewDelegate & UITableViewDataSource) -> Self {
        self.delegate = delegates
        self.dataSource = delegates
        return self
    }

    @discardableResult
    public func header<View: UIView>(_ view: View) -> View { tableHeaderView = view; return view }

    public var contentHeight: CGFloat {
        let lastFooterRect = rectForFooter(inSection: lastSection)
        return lastFooterRect.origin.y + lastFooterRect.size.height
    }

    public var lastSection: Int {
        var lastSection = numberOfSections - 1
        while lastSection > 0 && numberOfRows(inSection: lastSection) <= 0 {
            lastSection -= 1
        }
        return lastSection
    }
}
